src/Elastic.Apm.AspNetCore/ApmMiddleware.cs

```diff
@@ -42,21 +42,38 @@ public ApmMiddleware(RequestDelegate next, Tracer tracer, IApmAgent agent)
 
 		public async Task InvokeAsync(HttpContext context)
 		{
-			var transaction = StartTransaction(context);
+			var transaction = await StartTransactionAsync(context);
 
 			try
 			{
+				// We need to reset the current transaction (which is async local), since the StartTransactionAsync call above ended with all its
+				// child spans at this point, therefore need to set the current transaction here again.
+				_tracer.CurrentExecutionSegmentsContainer.CurrentTransaction = transaction;
 				await _next(context);
 			}
-			catch (Exception e) when (transaction != null && ExceptionFilter.Capture(e, transaction, context, _logger))
-			{ }
+			catch (Exception e) when (transaction != null)
+			{
+				transaction.CaptureException(e);
+				// It'd be nice to have this in an exception filter, but that would force us capturing the request body synchronously.
+				// Therefore we rather unwind the stack in the catch block and call the async method.
+				if (context != null && transaction.IsCaptureRequestBodyEnabled(true))

```

Comment:
Isn't call to `IsCaptureRequestBodyEnabled` redundant here considering that `CollectRequestBody` calls `IsCaptureRequestBodyEnabled` among other checks before actually reading the body?

Reply:
Right, yes, it is - (after I merged your change :) ) 👍 

I removed these checks.